include/openrave/kinbody.h

```diff
@@ -3696,7 +3699,16 @@ class OPENRAVE_API KinBody : public InterfaceBase
     std::vector<int> _vDOFIndices; ///< cached start joint indices, indexed by dof indices
     std::vector<uint64_t> _vLinkEnableStatesMask; /// bit mask containing enabled info of links. If bit 0 of _vLinkEnableBitMap[0] is 1, link 0 is enabled. If bit 2 of _vLinkEnableBitMap[1] is 0, link 66 is disabled. Used for fast access to the LinkInfo::_bIsEnabled
 
-    std::vector<std::pair<int16_t,int16_t> > _vAllPairsShortestPaths; ///< all-pairs shortest paths through the link hierarchy. The first value describes the parent link index, and the second value is an index into _vecjoints or _vPassiveJoints. If the second value is greater or equal to  _vecjoints.size() then it indexes into _vPassiveJoints.
+    /// All-pairs shortest paths through the link hierarchy.
+    /// The first value describes the parent link index, and the second value is an index into _vecjoints or _vPassiveJoints.
+    /// If the second value is greater or equal to  _vecjoints.size() then it indexes into _vPassiveJoints.
+    /// This value and its associated valid flag are mutable, as they are lazily calculated and so may need to be generated as part of a const method call.
+    mutable std::vector<std::pair<int16_t,int16_t> > _vAllPairsShortestPaths;
+
+    /// Flag to indicate whether the _vAllPairsShortestPaths has been calculated and safely accessed.
+    /// This value is mutable, as the associated data are lazily calculated and so this state may need to change as part of a const method call.
+    mutable bool _vAllPairsShortestPathsValid = false;
```

Comment:
@rschlaikjer  why not use `_vAllPairsShortestPaths.empty()` for determining if it needs computation instead of introducing a new flag `_vAllPairsShortestPathsValid`?

Reply:
I added the flag for the case where a body has no links - the vector will always be empty even if it's initialized, so there's no reason to spin through the rest of the function

Reply:
If there's no links, so what if it runs the function every time?

Reply:
fair enough; changed

Reply:
Great, thanks. This change requires MINOR version update since ABI changes